metadeploy/api/views.py

```diff
@@ -176,52 +191,160 @@ def filter_get_one(self, qs):
+        if scratch_org is None:
+            return Response("", status=status.HTTP_404_NOT_FOUND)
+        serializer = ScratchOrgJobSerializer(instance=scratch_org)
+        return Response(serializer.data)
+
+    def scratch_org_post(self, request, pk=None):
+        devhub_enabled = settings.DEVHUB_USERNAME
+        if not devhub_enabled:
+            return Response(status=status.HTTP_501_NOT_IMPLEMENTED)
+
+        plan = self.get_object()
+        valid_plan_type = (
+            plan.supported_orgs == SUPPORTED_ORG_TYPES.Scratch
+            or plan.supported_orgs == SUPPORTED_ORG_TYPES.Both
+        )
+        if not valid_plan_type:
+            return Response(
+                {"detail": "This plan does not support creating a scratch org."},
+                status=status.HTTP_409_CONFLICT,
+            )
+
+        new_data = request.data.copy()
+        new_data["plan"] = str(plan.pk)
+        serializer = ScratchOrgJobSerializer(data=new_data)
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        # Check queue status
+        # If overfull, return 503
+        queue = django_rq.get_queue("default")
+        if len(queue) > settings.MAX_QUEUE_LENGTH:
+            return Response(
+                {"detail": "Queue is overfull. Try again later."},
+                status=status.HTTP_503_SERVICE_UNAVAILABLE,
+            )
+
+        uuid = str(uuid4())
+        scratch_org = serializer.save(uuid=uuid)
+        request.session["scratch_org_id"] = uuid
+
+        serializer = ScratchOrgJobSerializer(instance=scratch_org)
+        return Response(
+            serializer.data,
+            status=status.HTTP_202_ACCEPTED,
+        )
+
+    @action(detail=True, methods=["post", "get"], permission_classes=(AllowAny,))
+    def scratch_org(self, request, pk=None):
+        if request.method == "GET":
+            return self.scratch_org_get(request)
+        if request.method == "POST":
+            return self.scratch_org_post(request)
+
 
 class OrgViewSet(viewsets.ViewSet):
+    permission_classes = (AllowAny,)
+
     def list(self, request):
         """
         This will return data on the user's current org. It is not a
         list endpoint, but does not take a pk, so we have to implement
         it this way.
         """
-        current_job = Job.objects.filter(
-            org_id=request.user.org_id, status=Job.Status.started
-        ).first()
-        current_preflight = PreflightResult.objects.filter(
-            org_id=request.user.org_id, status=PreflightResult.Status.started
-        ).first()
-        serializer = OrgSerializer(
-            {"current_job": current_job, "current_preflight": current_preflight}
-        )
-        return Response(serializer.data)
+        # TODO: make this return { anonymous: {}, nonymous: {} }
+        # TODO: make the ScratchOrgJob:uuid channel get events from
+        # related Jobs and PreflightResults
```

Comment:
@wlonk I'm not sure this is necessary, if we use `org_id` as the identifier for ScratchOrgs instead of `uuid`. That way the front end will subscribe to WS channels for scratch orgs and persistent orgs the same way (`model: 'org', id: <org_id>`) and the `ORG_CHANGED` events should work for both without changes.

Reply:
So we can't _only_ use the `org_id`; we don't have it at Scratch Org creation time, only when the job runs. So we have to have something else we can return to the user so they can sub to events on that object. That's why the UUID. Can we bootstrap from the UUID to the `org_id` when we get it? Is that reasonable? Is that terrible? I don't know.

Reply:
I don't think we _want_ to subscribe to this OrgViewSet WS channel until the Scratch Org is actually created -- so the front end can wait for the final `SCRATCH_ORG_CREATED` before doing that, and it could still use the `org_id`. I understand we need to be able to lookup ScratchOrgs by `uuid`, but I think for most cases (this included) we can still use `org_id`.

The scratch org model channel itself is another matter. Right now it's using `model: "scratch_org", id: <job_id>`. That works, but I'd propose changing it to `model: "scratch_org", id: <id>` so that it's more in line with our other patterns.